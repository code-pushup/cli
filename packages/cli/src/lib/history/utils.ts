import { HistoryOptions } from '@code-pushup/core';
import { getHashFromTag, isSemver } from '@code-pushup/utils';
import { HistoryCliOptions } from './history.model';

export async function normalizeHashOptions(
  processArgs: HistoryCliOptions & HistoryOptions,
): Promise<HistoryCliOptions & HistoryOptions> {
  const {
    onlySemverTags,
    // overwritten
    maxCount,
    ...opt
  } = processArgs;

  // eslint-disable-next-line functional/no-let, prefer-const
  let { from, to, ...processOptions } = opt;
  // if no semver filter is used resolve hash of tags, as hashes are used to collect history
  if (!onlySemverTags) {
    if (from && isSemver(from)) {
      const { hash } = await getHashFromTag(from);
      from = hash;
    }
    if (to && isSemver(to)) {
      const { hash } = await getHashFromTag(to);
      to = hash;
    }
  }

  return {
    ...processOptions,
    onlySemverTags,
    maxCount: maxCount && maxCount > 0 ? maxCount : undefined,
    from,
    to,
  };
}
