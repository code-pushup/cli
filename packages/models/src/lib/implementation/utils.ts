import type { CategoryConfig } from '../category-config';
import type { PluginConfig } from '../plugin-config';
import type { PluginReport } from '../report';

/**
 * Regular expression to validate a slug for categories, plugins and audits.
 * - audit (e.g. 'max-lines')
 * - category (e.g. 'performance')
 * Also validates ``and ` `
 */
export const slugRegex = /^[a-z\d]+(?:-[a-z\d]+)*$/;

/**
 * Regular expression to validate a filename.
 */
export const filenameRegex = /^(?!.*[ \\/:*?"<>|]).+$/;

/**
 * helper function to validate string arrays
 *
 * @param strings
 */
export function hasDuplicateStrings(strings: string[]): string[] | false {
  const sortedStrings = [...strings].sort();
  const duplStrings = sortedStrings.filter(
    (item, index) => index !== 0 && item === sortedStrings[index - 1],
  );

  return duplStrings.length === 0 ? false : [...new Set(duplStrings)];
}

/**
 * helper function to validate string arrays
 *
 * @param toCheck
 * @param existing
 */
export function hasMissingStrings(
  toCheck: string[],
  existing: string[],
): string[] | false {
  const nonExisting = toCheck.filter(s => !existing.includes(s));
  return nonExisting.length === 0 ? false : nonExisting;
}

/**
 * helper for error items
 */
export function errorItems(
  items: string[] | false,
  transform: (itemArr: string[]) => string = itemArr => itemArr.join(', '),
): string {
  return transform(items || []);
}

export function exists<T>(value: T): value is NonNullable<T> {
  return value != null;
}

/**
 * Get category references that do not point to any audit or group
 * @param categories
 * @param plugins
 * @returns Array of missing references.
 */
export function getMissingRefsForCategories(
  categories: CategoryConfig[],
  plugins: PluginConfig[] | PluginReport[],
) {
  if (categories.length === 0) {
    return false;
  }

  const auditRefsFromCategory = categories.flatMap(({ refs }) =>
    refs
      .filter(({ type }) => type === 'audit')
      .map(({ plugin, slug }) => `${plugin}/${slug}`),
  );
  const auditRefsFromPlugins = plugins.flatMap(({ audits, slug: pluginSlug }) =>
    audits.map(({ slug }) => `${pluginSlug}/${slug}`),
  );
  const missingAuditRefs = hasMissingStrings(
    auditRefsFromCategory,
    auditRefsFromPlugins,
  );

  const groupRefsFromCategory = categories.flatMap(({ refs }) =>
    refs
      .filter(({ type }) => type === 'group')
      .map(({ plugin, slug }) => `${plugin}#${slug} (group)`),
  );
  const groupRefsFromPlugins = plugins.flatMap(({ groups, slug: pluginSlug }) =>
    Array.isArray(groups)
      ? groups.map(({ slug }) => `${pluginSlug}#${slug} (group)`)
      : [],
  );
  const missingGroupRefs = hasMissingStrings(
    groupRefsFromCategory,
    groupRefsFromPlugins,
  );

  const missingRefs = [missingAuditRefs, missingGroupRefs]
    .filter((refs): refs is string[] => Array.isArray(refs) && refs.length > 0)
    .flat();

  return missingRefs.length > 0 ? missingRefs : false;
}

export function missingRefsForCategoriesErrorMsg(
  categories: CategoryConfig[],
  plugins: PluginConfig[] | PluginReport[],
) {
  const missingRefs = getMissingRefsForCategories(categories, plugins);
  return `The following category references need to point to an audit or group: ${errorItems(
    missingRefs,
  )}`;
}
