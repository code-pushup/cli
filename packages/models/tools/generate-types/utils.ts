import { mkdir, stat, writeFile } from 'node:fs/promises';
import { dirname } from 'node:path';
import { ZodTypeAny } from 'zod';
import { createTypeAlias, printNode, zodToTs } from 'zod-to-ts';

export function generateTypesString(
  schemas: Record<string, ZodTypeAny>,
): string {
  return Object.entries(schemas)
    .map(([identifier, schema]) => {
      const typeName = schemaNameToTypeName(identifier);
      const { node } = zodToTs(schema, typeName);
      const typeAlias = createTypeAlias(node, typeName);
      return `export ${printNode(typeAlias)}`;
    })
    .join('\n');
}

export function schemaNameToTypeName(schemaName: string): string {
  const name = schemaName.replace(/Schema/, '');
  return name[0].toUpperCase() + name.slice(1);
}

export function generateMdFile(typesString: string): string {
  return [
    '# Code PushUp config file reference',
    'The `code-pushup.config.(ts|mjs|js)` file should conform to the following type definition:',
    '```ts',
    typesString,
    '```',
  ].join('\n');
}

export async function safeWriteFile(path: string, content: string) {
  const dir = dirname(path) as string;
  const stats = await stat(dir);
  if (!stats.isDirectory()) {
    await mkdir(dir);
  }
  await writeFile(path, content, { encoding: 'utf8' });
}
