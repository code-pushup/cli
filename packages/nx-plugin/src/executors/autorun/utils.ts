import { globalConfig, persistConfig, uploadConfig } from '../internal/config';
import { NormalizedExecutorContext } from '../internal/context';
import {
  AutorunCommandExecutorOnlyOptions,
  AutorunCommandExecutorOptions,
} from './schema';

export function parseAutorunExecutorOnlyOptions(
  options: Partial<AutorunCommandExecutorOnlyOptions>,
): AutorunCommandExecutorOnlyOptions {
  const { projectPrefix, dryRun, onlyPlugins } = options;
  return {
    ...(projectPrefix && { projectPrefix }),
    ...(dryRun != null && { dryRun }),
    ...(onlyPlugins && { onlyPlugins }),
  };
}

export function parseAutorunExecutorOptions(
  options: Partial<AutorunCommandExecutorOptions>,
  normalizedContext: NormalizedExecutorContext,
): AutorunCommandExecutorOptions {
  const { projectPrefix, persist, upload } = options;
  return {
    ...parseAutorunExecutorOnlyOptions(options),
    ...globalConfig(options, normalizedContext),
    persist: persistConfig({ projectPrefix, ...persist }, normalizedContext),
    upload: uploadConfig({ projectPrefix, ...upload }, normalizedContext),
  };
}
