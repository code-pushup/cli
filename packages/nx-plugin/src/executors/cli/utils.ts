import {
  globalConfig,
  persistConfig,
  uploadConfig,
} from '../internal/config.js';
import type { NormalizedExecutorContext } from '../internal/context.js';
import type {
  AutorunCommandExecutorOnlyOptions,
  AutorunCommandExecutorOptions,
} from './schema.js';

export function parseAutorunExecutorOnlyOptions(
  options: Partial<AutorunCommandExecutorOnlyOptions>,
): AutorunCommandExecutorOnlyOptions {
  const { projectPrefix, dryRun, onlyPlugins } = options;
  return {
    ...(projectPrefix && { projectPrefix }),
    ...(dryRun != null && { dryRun }),
    ...(onlyPlugins && { onlyPlugins }),
  };
}

export function parseAutorunExecutorOptions(
  options: Partial<AutorunCommandExecutorOptions>,
  normalizedContext: NormalizedExecutorContext,
): AutorunCommandExecutorOptions {
  const { projectPrefix, persist, upload, command } = options;
  const needsUploadParams =
    command === 'upload' || command === 'autorun' || command === undefined;
  const uploadCfg = uploadConfig(
    { projectPrefix, ...upload },
    normalizedContext,
  );
  const hasApiToken = uploadCfg?.apiKey != null;
  return {
    ...parseAutorunExecutorOnlyOptions(options),
    ...globalConfig(options, normalizedContext),
    persist: persistConfig({ projectPrefix, ...persist }, normalizedContext),
    // @TODO This is a hack to avoid validation errors of upload config for commands that dont need it.
    // Fix: use utils and execute the core logic directly
    // Blocked by Nx plugins can't compile to es6
    ...(needsUploadParams && hasApiToken ? { upload: uploadCfg } : {}),
  };
}

/**
 * Deeply merges executor options.
 *
 * @param targetOptions - The original options from the target configuration.
 * @param cliOptions - The options from Nx, combining target options and CLI arguments.
 * @returns A new object with deeply merged properties.
 *
 * Nx performs a shallow merge by default, where command-line arguments can override entire objects
 * (e.g., `--persist.filename` replaces the entire `persist` object).
 * This function ensures that nested properties are deeply merged,
 * preserving the original target options where CLI arguments are not provided.
 */
export function mergeExecutorOptions(
  targetOptions: Partial<AutorunCommandExecutorOptions>,
  cliOptions: Partial<AutorunCommandExecutorOptions>,
): AutorunCommandExecutorOptions {
  return {
    ...targetOptions,
    ...cliOptions,
    ...(targetOptions?.persist || cliOptions?.persist
      ? {
          persist: {
            ...targetOptions?.persist,
            ...cliOptions?.persist,
          },
        }
      : {}),
    ...(targetOptions?.upload || cliOptions?.upload
      ? {
          upload: {
            ...targetOptions?.upload,
            ...cliOptions?.upload,
          },
        }
      : {}),
  };
}
