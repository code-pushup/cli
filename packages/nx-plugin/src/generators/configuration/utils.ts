import type { ExtractArrays } from '@code-pushup/utils';
import type { ExecutableCode } from './types';

export function normalizeExecutableCode(
  executableCode: ExecutableCode,
): ExtractArrays<ExecutableCode> {
  const { fileImports: rawFileImports, codeStrings: rawCodeStrings } =
    executableCode;

  return {
    fileImports: normalizeItemOrArray(rawFileImports),
    codeStrings: normalizeItemOrArray(rawCodeStrings),
  };
}

export function normalizeItemOrArray<T>(itemOrArray: T | T[]): T[];
export function normalizeItemOrArray<T>(
  itemOrArray: T | T[] | undefined,
): T[] | undefined {
  if (itemOrArray == null) {
    return undefined;
  }
  if (Array.isArray(itemOrArray)) {
    return itemOrArray;
  }
  return [itemOrArray];
}

// Return a formatted JSON in TS object with the same keys as the input object but remove the " for the properties
export function formatObjectToFormattedJsString(
  jsonObj?:
    | {
        [key: string]: unknown;
      }
    | Array<unknown>,
): string | undefined {
  if (!jsonObj) {
    return;
  }
  // Convert JSON object to a string with indentation
  const jsonString = JSON.stringify(jsonObj, null, 2);

  // Remove double quotes around property names
  return jsonString.replace(/"(\w+)":/g, '$1:');
}

export function formatArrayToLinesOfJsString(
  lines?: string[],
  separator = '\n',
) {
  if (lines == null || lines.length === 0) {
    return;
  }
  return lines.join(separator).replace(/'/g, '"');
}

export function formatArrayToJSArray(lines?: string[]) {
  if (!Array.isArray(lines)) {
    return;
  }

  return `[${formatArrayToLinesOfJsString(lines, ',\n') ?? ''}]`.replace(
    /"/g,
    '',
  );
}
