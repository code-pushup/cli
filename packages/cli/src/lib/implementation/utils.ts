import { toArray } from '@code-pushup/utils';

// log error and flush stdout so that Yargs doesn't suppress it
// related issue: https://github.com/yargs/yargs/issues/2118
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function logErrorBeforeThrow<T extends (...args: any[]) => any>(
  fn: T,
): T {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return (async (...args: any[]) => {
    try {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-argument
      return await fn(...args);
    } catch (error) {
      console.error(error);
      await new Promise(resolve => process.stdout.write('', resolve));
      throw error;
    }
  }) as T;
}

export function coerceArray<T extends string>(param: T | T[] = []): T[] {
  return [
    ...new Set(toArray(param).flatMap((f: T) => f.split(',') as T[]) || []),
  ];
}
