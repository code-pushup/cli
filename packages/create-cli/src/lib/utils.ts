import { readFile, rm, stat, writeFile } from 'node:fs/promises';
import { CODE_PUSHUP_UNICODE_LOGO } from '@code-pushup/utils';
import {
  NX_JSON_CONTENT,
  NX_JSON_FILENAME,
  PROJECT_JSON_CONTENT,
  PROJECT_JSON_FILENAME,
  PROJECT_NAME,
} from './constants.js';

export type SetupResult = {
  filename: string;
  teardown: boolean;
};

export async function setupFile(
  filename: string,
  content = '',
): Promise<SetupResult> {
  const setupResult: SetupResult = {
    filename,
    teardown: false,
  };

  try {
    const stats = await stat(filename);
    if (!stats.isFile()) {
      await writeFile(filename, content);
    }
  } catch (error) {
    if (
      error instanceof Error &&
      error.message.includes('no such file or directory')
    ) {
      await writeFile(filename, content);
      return {
        ...setupResult,
        teardown: true,
      };
    } else {
      console.error(error);
    }
  }

  return setupResult;
}

export function parseNxProcessOutput(output: string) {
  return output.trim().replace('NX', CODE_PUSHUP_UNICODE_LOGO);
}

export async function setupNxContext(): Promise<{
  nxJsonTeardown: boolean;
  projectJsonTeardown: boolean;
  projectName: string;
}> {
  const { teardown: nxJsonTeardown } = await setupFile(
    NX_JSON_FILENAME,
    NX_JSON_CONTENT,
  );
  const { teardown: projectJsonTeardown } = await setupFile(
    PROJECT_JSON_FILENAME,
    PROJECT_JSON_CONTENT,
  );

  const projectJsonContent = await readFile(PROJECT_JSON_FILENAME, 'utf8');
  const { name = PROJECT_NAME } = JSON.parse(projectJsonContent) as {
    name: string;
  };

  return {
    nxJsonTeardown,
    projectJsonTeardown,
    projectName: name,
  };
}

export async function teardownNxContext({
  nxJsonTeardown,
  projectJsonTeardown,
}: {
  nxJsonTeardown: boolean;
  projectJsonTeardown: boolean;
}) {
  const filesToDelete = [
    ...(nxJsonTeardown ? [NX_JSON_FILENAME] : []),
    ...(projectJsonTeardown ? [PROJECT_JSON_FILENAME] : []),
  ];
  await Promise.all(filesToDelete.map(file => rm(file)));
}
