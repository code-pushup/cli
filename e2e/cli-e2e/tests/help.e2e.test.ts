import { executeProcess } from '@code-pushup/utils';

describe('CLI help', () => {
  it('should print help with help command', async () => {
    const { code, stdout, stderr } = await executeProcess({
      command: 'code-pushup',
      args: ['help'],
    });
    expect(code).toBe(0);
    expect(stderr).toBe('');
    expect(stdout).toMatchSnapshot();
  });

  it('should produce the same output to stdout for both help argument and help command', async () => {
    const helpArgResult = await executeProcess({
      command: 'code-pushup',
      args: ['help'],
    });
    const helpCommandResult = await executeProcess({
      command: 'code-pushup',
      args: ['--help'],
    });
    expect(helpArgResult.code).toBe(0);
    expect(helpCommandResult.code).toBe(0);
    expect(helpArgResult.stdout).toBe(helpCommandResult.stdout);
  });
});
