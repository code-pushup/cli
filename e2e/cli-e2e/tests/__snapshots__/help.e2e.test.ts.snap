// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CLI help > should print help with help command 1`] = `
"[1mCode PushUp CLI[22m

Commands:
  code-pushup               Shortcut for running collect followed by upload
                                                                       [default]
  code-pushup autorun       Shortcut for running collect followed by upload
  code-pushup collect       Run Plugins and collect results
  code-pushup upload        Upload report results to the portal
  code-pushup history       Collect reports for commit history
  code-pushup compare       Compare 2 report files and create a diff file
  code-pushup print-config  Print config

Global Options:
      --progress     Show progress bar in stdout.      [boolean] [default: true]
      --verbose      When true creates more verbose output. This is helpful when
                      debugging.                      [boolean] [default: false]
      --config       Path to config file, e.g. code-pushup.config.ts. By default
                      it loads code-pushup.config.(ts|mjs|js).          [string]
      --tsconfig     Path to a TypeScript config, to be used when loading config
                      file.                                             [string]
      --onlyPlugins  List of plugins to run. If not set all plugins are run.
                                                           [array] [default: []]

Persist Options:
      --persist.outputDir  Directory for the produced reports           [string]
      --persist.filename   Filename for the produced reports.           [string]
      --persist.format     Format of the report output. e.g. \`md\`, \`json\`[array]

Upload Options:
      --upload.organization  Organization slug from portal              [string]
      --upload.project       Project slug from portal                   [string]
      --upload.server        URL to your portal server                  [string]
      --upload.apiKey        API key for the portal server              [string]

Options:
  -h, --help  Show help                                                [boolean]

Examples:
  code-pushup                               Run collect followed by upload based
                                             on configuration from code-pushup.c
                                            onfig.* file.
  code-pushup collect --tsconfig=tsconfig.  Run collect using custom tsconfig to
  base.json                                  parse code-pushup.config.ts file.
  code-pushup collect --onlyPlugins=covera  Run collect with only coverage plugi
  ge                                        n, other plugins from config file wi
                                            ll be skipped.
  code-pushup upload --persist.outputDir=d  Upload dist/cp-report.json to portal
  ist --persist.filename=cp-report --uploa   using API key from environment vari
  d.apiKey=$CP_API_KEY                      able
  code-pushup print-config --config code-p  Print resolved config object parsed
  ushup.config.test.js                      from custom config location
"
`;
